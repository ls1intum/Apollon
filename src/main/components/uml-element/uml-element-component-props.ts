export type UMLElementComponentProps = {
  id: string;
  children?: React.ReactNode;
};
