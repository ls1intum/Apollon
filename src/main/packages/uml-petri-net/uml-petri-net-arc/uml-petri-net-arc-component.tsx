import React, { SFC } from 'react';
import { Point } from '../../../utils/geometry/point';
import { UMLPetriNetArc } from './uml-petri-net-arc';

export const UMLPetriNetArcComponent: SFC<Props> = ({ element }) => {
  const [start, end] = element.path.map((p) => new Point(p.x, p.y));
  const line = end.subtract(start);
  const norm = line.normalize();
  const center = start.add(norm.scale(0.5 * line.length));

  const displayMultiplicity = element.name !== UMLPetriNetArc.defaultMultiplicity;

  return (
    <g>
      <marker
        id={`marker-${element.id}`}
        viewBox="0 0 30 30"
        markerWidth="22"
        markerHeight="30"
        refX="30"
        refY="15"
        orient="auto"
        markerUnits="strokeWidth"
      >
        <path d="M0,1 L0,29 L30,15 z" fill="black" />
      </marker>
      <path
        id={`textpath-${element.id}`}
        d={`
        M ${start.x} ${start.y - 10}
        L ${end.x} ${end.y - 10}
    `}
      />
      {displayMultiplicity && (
        <text
          dominantBaseline="middle"
          textAnchor="middle"
          fontWeight="bold"
          transform={
            norm.x < 0
              ? `
            translate(${center.x}, ${center.y})
            rotate(180)
            translate(${-center.x}, ${-center.y})
          `
              : undefined
          }
          pointerEvents="none"
        >
          <textPath xlinkHref={`#textpath-${element.id}`} startOffset="50%">
            {element.name}
          </textPath>
        </text>
      )}
      <polyline
        points={element.path.map((point) => `${point.x} ${point.y}`).join(',')}
        stroke="black"
        fill="none"
        strokeWidth={1}
        markerEnd={`url(#marker-${element.id})`}
      />
    </g>
  );
};

export interface Props {
  element: UMLPetriNetArc;
}
