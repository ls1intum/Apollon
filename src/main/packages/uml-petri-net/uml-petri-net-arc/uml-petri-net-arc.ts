import { PetriNetRelationshipType } from '../index';
import { IUMLRelationship, UMLRelationship } from '../../../services/uml-relationship/uml-relationship';
import { DeepPartial } from '../../../typings';

export class UMLPetriNetArc extends UMLRelationship {
  static features = { ...UMLRelationship.features, straight: true };
  static defaultMultiplicity = '1';
  type = PetriNetRelationshipType.PetriNetArc;
  name = UMLPetriNetArc.defaultMultiplicity;

  constructor(values?: DeepPartial<IUMLRelationship>) {
    super(values);
    this.name = (values && values.name) || this.name;
  }
}
