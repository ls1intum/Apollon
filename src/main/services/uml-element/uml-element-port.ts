export enum Direction {
  Up = 'Up',
  Right = 'Right',
  Down = 'Down',
  Left = 'Left',
  Upright = 'Upright',
  Upleft = 'Upleft',
  Downright = 'Downright',
  Downleft = 'Downleft',
  Topright = 'Topright',
  Topleft = 'Topleft',
  Bottomright = 'Bottomright',
  Bottomleft = 'Bottomleft',
}

export interface IUMLElementPort {
  element: string;
  direction: Direction;
}

export function getOppositeDirection(direction: Direction): Direction {
  switch (direction) {
    case Direction.Down:
      return Direction.Up;
    case Direction.Left:
      return Direction.Right;
    case Direction.Right:
      return Direction.Left;
    case Direction.Up:
      return Direction.Down;
    case Direction.Upright:
      return Direction.Upleft;
    case Direction.Downright:
      return Direction.Downleft;
    case Direction.Upleft:
      return Direction.Upright;
    case Direction.Downleft:
      return Direction.Downright;
    case Direction.Topright:
      return Direction.Topleft;
    case Direction.Bottomright:
      return Direction.Bottomleft;
    case Direction.Topleft:
      return Direction.Topright;
    case Direction.Bottomleft:
      return Direction.Bottomright;
    default:
      throw Error(`Could not determine opposite direction for direction of ${direction}`);
  }
}
