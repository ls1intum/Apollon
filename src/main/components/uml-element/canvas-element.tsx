import React, { Component, ComponentClass, SVGProps } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Components } from '../../packages/components';
import { UMLElementType } from '../../packages/uml-element-type';
import { ApollonView } from '../../services/editor/editor-types';
import { UMLContainer } from '../../services/uml-container/uml-container';
import { IUMLElement } from '../../services/uml-element/uml-element';
import { UMLElementRepository } from '../../services/uml-element/uml-element-repository';
import { ModelState } from '../store/model-state';
import { withTheme, withThemeProps } from '../theme/styles';
import { UMLElementComponentProps } from './uml-element-component-props';

const STROKE = 5;

type OwnProps = { child?: ComponentClass<UMLElementComponentProps> } & UMLElementComponentProps &
  SVGProps<SVGSVGElement>;

type StateProps = {
  hovered: boolean;
  selected: boolean;
  moving: boolean;
  interactive: boolean;
  interactable: boolean;
  element: IUMLElement;
};

type DispatchProps = {};

type Props = OwnProps & StateProps & DispatchProps & withThemeProps;

const enhance = compose<ComponentClass<OwnProps>>(
  withTheme,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(
    (state, props) => ({
      hovered: state.hovered[0] === props.id,
      selected: state.selected.includes(props.id),
      moving: state.moving.includes(props.id),
      interactive: state.interactive.includes(props.id),
      interactable: state.editor.view === ApollonView.Exporting || state.editor.view === ApollonView.Highlight,
      element: state.elements[props.id],
    }),
    {},
  ),
);

class CanvasElementComponent extends Component<Props> {
  render() {
    const {
      hovered,
      selected,
      moving,
      interactive,
      interactable,
      element,
      child: ChildComponent,
      children,
      theme,
      ...props
    } = this.props;

    let elements = null;
    if (UMLContainer.isUMLContainer(element) && ChildComponent) {
      elements = element.ownedElements.map((id) => <ChildComponent key={id} id={id} />);
    }
    const ElementComponent = Components[element.type as UMLElementType];

    const highlight =
      interactable && interactive
        ? theme.interactive.normal
        : interactable && hovered
        ? theme.interactive.hovered
        : element.highlight
        ? element.highlight
        : element.fillColor
        ? element.fillColor
        : theme.color.background;

    const selectedByList = element.selectedBy || [];
    return (
      <svg
        {...props}
        {...element.bounds}
        pointerEvents={moving ? 'none' : undefined}
        fillOpacity={moving ? 0.7 : undefined}
        fill={highlight}
      >
        <ElementComponent fillColor={highlight} element={UMLElementRepository.get(element)}>
          {elements}
        </ElementComponent>
        {children}
        {!interactable && (hovered || selected) && (
          <rect
            x={-STROKE / 2}
            y={-STROKE / 2}
            width={element.bounds.width + STROKE}
            height={element.bounds.height + STROKE}
            fill="none"
            stroke="#0064ff"
            strokeOpacity="0.2"
            strokeWidth={STROKE}
            pointerEvents="none"
          />
        )}
        {selectedByList.length > 0 && (
          <g>
            {selectedByList.map((selectedBy, index) => {
              const indicatorPosition = 'translate(' + (element.bounds.width + STROKE) + ' ' + index * 32 + ')';
              return (
                <g key={selectedBy.name + '_' + selectedBy.color} id={selectedBy.name + '_' + selectedBy.color}>
                  <rect
                    x={-STROKE / 2}
                    y={-STROKE / 2}
                    width={element.bounds.width + STROKE}
                    height={element.bounds.height + STROKE}
                    fill="none"
                    stroke={selectedBy.color}
                    strokeOpacity="0.2"
                    strokeWidth={STROKE}
                    pointerEvents="none"
                  />

                  <g transform={indicatorPosition} pointerEvents="none">
                    <rect
                      fillOpacity="0.2"
                      rx="10"
                      x="-40"
                      y="-20"
                      width="85px"
                      height="30px"
                      fill={selectedBy.color}
                    />
                    <text>
                      <tspan textAnchor="middle">
                        {selectedBy.name.length < 8 ? selectedBy.name : selectedBy.name.substring(0, 6) + '..'}
                      </tspan>
                    </text>
                  </g>
                </g>
              );
            })}
          </g>
        )}
      </svg>
    );
  }
}

export const CanvasElement = enhance(CanvasElementComponent);
