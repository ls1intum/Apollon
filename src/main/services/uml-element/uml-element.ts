import { DeepPartial } from 'redux';
import { UMLDiagramType } from '../../packages/diagram-type';
import { UMLElementType } from '../../packages/uml-element-type';
import { UMLRelationshipType } from '../../packages/uml-relationship-type';
import * as Apollon from '../../typings';
import { assign } from '../../utils/fx/assign';
import { IBoundary } from '../../utils/geometry/boundary';
import { Point } from '../../utils/geometry/point';
import { uuid } from '../../utils/uuid';
import { ILayer } from '../layouter/layer';
import { ILayoutable } from '../layouter/layoutable';
import { UMLElementFeatures } from './uml-element-features';
import { Direction } from './uml-element-port';

/** Interface of a `UMLElement` defining the properties persisted in the internal storage */
export interface IUMLElement {
  /** Unique Identifier of the `UMLElement` */
  id: string;
  /** Visual name of the `UMLElement` */
  name: string;
  /** Distinct type to recreate the `UMLElement` */
  type: UMLElementType | UMLRelationshipType | UMLDiagramType;
  /** Optional owner of the `UMLElement` */
  owner: string | null;
  /** Position and sizing of the `UMLElement` */
  bounds: IBoundary;
  /** Highlight the element with a specified color */
  highlight?: string;
  /** Colors of the element */
  fillColor?: string;
  strokeColor?: string;
  textColor?: string;
  /** Note to show for element's assessment */
  assessmentNote?: string;
}

export const enum ResizeFrom {
  TOPLEFT = 'topLeft',
  TOPRIGHT = 'topRight',
  BOTTOMLEFT = 'bottomLeft',
  BOTTOMRIGHT = 'bottomRight',
}

export const getPortsForElement = (element: IUMLElement): { [key in Direction]: Point } => {
  return {
    [Direction.Up]: new Point(element.bounds.width / 2, 0),
    [Direction.Right]: new Point(element.bounds.width, element.bounds.height / 2),
    [Direction.Down]: new Point(element.bounds.width / 2, element.bounds.height),
    [Direction.Left]: new Point(0, element.bounds.height / 2),
    [Direction.Upright]: new Point(element.bounds.width, element.bounds.height / 4),
    [Direction.Downright]: new Point(element.bounds.width, (3 * element.bounds.height) / 4),
    [Direction.Upleft]: new Point(0, element.bounds.height / 4),
    [Direction.Downleft]: new Point(0, (3 * element.bounds.height) / 4),
    [Direction.Topright]: new Point((3 * element.bounds.width) / 4, 0),
    [Direction.Bottomright]: new Point((3 * element.bounds.width) / 4, element.bounds.height),
    [Direction.Topleft]: new Point(element.bounds.width / 4, 0),
    [Direction.Bottomleft]: new Point(element.bounds.width / 4, element.bounds.height),
  };
};

/** Class implementation of `IUMLElement` to use inheritance at runtime */
export abstract class UMLElement implements IUMLElement, ILayoutable {
  /** `UMLElement` type specific feature flags */
  static features: UMLElementFeatures = {
    connectable: true,
    droppable: false,
    hoverable: true,
    movable: true,
    resizable: true,
    selectable: true,
    updatable: true,
    alternativePortVisualization: false,
  };

  static supportedRelationships: UMLRelationshipType[] = [];

  /** Checks whether an `IUMLElement` is of type `UMLElementType` */
  static isUMLElement = (element: IUMLElement): element is IUMLElement & { type: UMLElementType } =>
    element.type in UMLElementType;

  id = uuid();
  name = '';
  abstract type: UMLElementType | UMLRelationshipType | UMLDiagramType;
  bounds = { x: 0, y: 0, width: 200, height: 100 };
  owner = null as string | null;
  highlight?: string;
  fillColor?: string;
  strokeColor?: string;
  textColor?: string;
  assessmentNote?: string;
  resizeFrom: ResizeFrom = ResizeFrom.BOTTOMRIGHT;

  constructor(values?: DeepPartial<IUMLElement>) {
    assign<IUMLElement>(this, values);
  }

  /**
   * Clones an instance of `UMLElement`
   *
   * @param override - Override existing properties.
   */
  clone<T extends UMLElement>(override?: DeepPartial<IUMLElement>): T {
    const Constructor = this.constructor as new (values?: DeepPartial<IUMLElement>) => T;
    const values: IUMLElement = { ...this, ...override, id: uuid() };

    return new Constructor(values);
  }

  /** Serializes an `UMLElement` to an `Apollon.UMLElement` */
  serialize(children?: UMLElement[]): Apollon.UMLModelElement {
    return {
      id: this.id,
      name: this.name,
      type: this.type,
      owner: this.owner,
      bounds: this.bounds,
      highlight: this.highlight,
      fillColor: this.fillColor,
      strokeColor: this.strokeColor,
      textColor: this.textColor,
      assessmentNote: this.assessmentNote,
    };
  }

  /** Deserializes an `Apollon.UMLElement` to an `UMLElement` */
  deserialize<T extends Apollon.UMLModelElement>(values: T, children?: Apollon.UMLModelElement[]) {
    this.id = values.id;
    this.name = values.name;
    this.type = values.type;
    this.owner = values.owner || null;
    this.bounds = { ...values.bounds };
    this.highlight = values.highlight;
    this.fillColor = values.fillColor;
    this.strokeColor = values.strokeColor;
    this.textColor = values.textColor;
    this.assessmentNote = values.assessmentNote;
  }

  abstract render(canvas: ILayer): ILayoutable[];
}
