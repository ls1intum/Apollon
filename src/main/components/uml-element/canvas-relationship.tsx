import React, { Component, ComponentClass, SVGProps } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Components } from '../../packages/components';
import { UMLRelationshipType } from '../../packages/uml-relationship-type';
import { ApollonMode, ApollonView } from '../../services/editor/editor-types';
import { Direction } from '../../services/uml-element/uml-element-port';
import { IUMLRelationship } from '../../services/uml-relationship/uml-relationship';
import { UMLRelationshipRepository } from '../../services/uml-relationship/uml-relationship-repository';
import { AsyncDispatch } from '../../utils/actions/actions';
import { Point } from '../../utils/geometry/point';
import { getClientEventCoordinates } from '../../utils/touch-event';
import { ModelState } from '../store/model-state';
import { withTheme, withThemeProps } from '../theme/styles';
import { UMLElementComponentProps } from './uml-element-component-props';

type OwnProps = UMLElementComponentProps & SVGProps<SVGSVGElement>;

type StateProps = {
  hovered: boolean;
  selected: boolean;
  interactive: boolean;
  interactable: boolean;
  reconnecting: boolean;
  disabled: boolean;
  relationship: IUMLRelationship;
  mode: ApollonMode;
  scale: number;
};

type DispatchProps = {
  startwaypointslayout: AsyncDispatch<typeof UMLRelationshipRepository.startWaypointsLayout>;
  endwaypointslayout: AsyncDispatch<typeof UMLRelationshipRepository.endWaypointsLayout>;
};

type Props = OwnProps & StateProps & DispatchProps & withThemeProps;

const initialState = {
  offset: new Point(),
  handlerIndex: 0,
  path: [
    {
      x: 0,
      y: 0,
    },
  ],
};

type State = typeof initialState;

const enhance = compose<ComponentClass<OwnProps>>(
  withTheme,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(
    (state, props) => ({
      hovered: state.hovered[0] === props.id,
      selected: state.selected.includes(props.id),
      interactive: state.interactive.includes(props.id),
      interactable: state.editor.view === ApollonView.Exporting || state.editor.view === ApollonView.Highlight,
      reconnecting: !!state.reconnecting[props.id],
      disabled: !!Object.keys(state.reconnecting).length || !!Object.keys(state.connecting).length,
      relationship: state.elements[props.id] as IUMLRelationship,
      mode: state.editor.mode as ApollonMode,
      scale: state.editor.scale || 1.0,
    }),
    {
      startwaypointslayout: UMLRelationshipRepository.startWaypointsLayout,
      endwaypointslayout: UMLRelationshipRepository.endWaypointsLayout,
    },
  ),
);

export class CanvasRelationshipComponent extends Component<Props, State> {
  state = initialState;

  render() {
    const {
      hovered,
      selected,
      interactive,
      interactable,
      reconnecting,
      disabled,
      relationship,
      children,
      theme,
      mode,
      scale,
      startwaypointslayout,
      endwaypointslayout,
      ...props
    } = this.props;

    // increase relationship hit box in assessment mode
    const STROKE = mode === ApollonMode.Assessment ? 35 : 15;

    const ChildComponent = Components[relationship.type as UMLRelationshipType];

    const points = relationship.path.map((point) => `${point.x} ${point.y}`).join(',');

    const midPoints: { mpX: number; mpY: number }[] = [];
    relationship.path.map((point, index) => {
      const mpX = (relationship.path[index].x + relationship.path[index + 1]?.x) / 2;
      const mpY = (relationship.path[index].y + relationship.path[index + 1]?.y) / 2;
      if (!isNaN(mpX) && !isNaN(mpY)) midPoints.push({ mpX, mpY });
    });

    midPoints.pop();
    midPoints.shift();

    const highlight =
      interactable && interactive
        ? theme.interactive.normal
        : interactable && hovered
        ? theme.interactive.hovered
        : hovered || selected
        ? 'rgba(0, 100, 255, 0.2)'
        : relationship.highlight
        ? relationship.highlight
        : 'rgba(0, 100, 255, 0)';

    return (
      <svg
        {...props}
        {...relationship.bounds}
        visibility={reconnecting ? 'hidden' : undefined}
        pointerEvents={disabled ? 'none' : 'stroke'}
      >
        <polyline points={points} stroke={highlight} fill="none" strokeWidth={STROKE} />
        <ChildComponent scale={scale} element={UMLRelationshipRepository.get(relationship)} />
        {children}
        {midPoints.map((point, index) => {
          return (
            <circle
              visibility={interactive || interactable ? 'hidden' : undefined}
              pointerEvents={interactive || interactable ? 'none' : 'all'}
              style={{ cursor: 'grab' }}
              key={props.id + '_' + point.mpX + '_' + point.mpY}
              cx={point.mpX}
              cy={point.mpY}
              r="15"
              onPointerDown={(e) => {
                this.onPointerDown(e, index, point);
              }}
              fill={highlight}
            />
          );
        })}
      </svg>
    );
  }

  onPointerDown = (event: any, handlerIndex: number, point: { mpX: number; mpY: number }) => {
    this.setState({ handlerIndex, offset: new Point(event.clientX - point.mpX, event.clientY - point.mpY) });
    document.addEventListener('pointermove', this.onPointerMove);
    document.addEventListener('pointerup', this.onPointerUp, { once: true });
  };

  onPointerMove = (event: any) => {
    const handlerIndex = this.state.handlerIndex;
    const waypointDirection = handlerIndex % 2 ? 'horizontal' : 'vertical';

    const clientEventCoordinates = getClientEventCoordinates(event);
    const x = clientEventCoordinates.clientX - this.state.offset.x;
    const y = clientEventCoordinates.clientY - this.state.offset.y;

    // Update relationship points here
    this.updateRelationshipPoints(waypointDirection, handlerIndex, x, y);
  };

  onPointerUp = (event: any) => {
    this.props.endwaypointslayout(this.props.id);
    const element = event.currentTarget;
    element.removeEventListener('pointermove', this.onPointerMove);
  };

  updateRelationshipPoints = (waypointDirection: string, handlerIndex: number, x: number, y: number) => {
    const startPoint = handlerIndex + 1;
    const endPoint = Number(startPoint) + 1;
    const sourceDirection = this.props.relationship.source.direction;

    switch (waypointDirection) {
      case 'horizontal':
        sourceDirection === Direction.Up || sourceDirection === Direction.Down
          ? this.updateXCoordinate(startPoint, endPoint, x, y)
          : this.updateYCoordinate(startPoint, endPoint, x, y);
        break;

      case 'vertical':
        sourceDirection === Direction.Up || sourceDirection === Direction.Down
          ? this.updateYCoordinate(startPoint, endPoint, x, y)
          : this.updateXCoordinate(startPoint, endPoint, x, y);
        break;

      default:
        break;
    }

    this.setState({ path: this.props.relationship.path });
    this.props.startwaypointslayout(this.props.id, this.props.relationship.path);
  };

  updateXCoordinate = (startPoint: number, endPoint: number, x: number, y: number) => {
    this.props.relationship.path[startPoint].x = x;
    this.props.relationship.path[endPoint].x = x;
  };

  updateYCoordinate = (startPoint: number, endPoint: number, x: number, y: number) => {
    this.props.relationship.path[startPoint].y = y;
    this.props.relationship.path[endPoint].y = y;
  };
}

export const CanvasRelationship = enhance(CanvasRelationshipComponent);
