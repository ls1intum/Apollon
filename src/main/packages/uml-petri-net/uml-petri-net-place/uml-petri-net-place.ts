import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { IUMLElement, UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementFeatures } from '../../../services/uml-element/uml-element-features';
import { IBoundary } from '../../../utils/geometry/boundary';
import { UMLElementType } from '../../uml-element-type';
import { DeepPartial } from 'redux';
import * as Apollon from '../../../typings';
import { PetriNetElementType } from '../index';
import { assign } from '../../../utils/fx/assign';

export class UMLPetriNetPlace extends UMLElement {
  static features: UMLElementFeatures = { ...UMLElement.features, resizable: false };
  static defaultCapacity = Number.POSITIVE_INFINITY;

  // currently we need to add this, because otherwise this will be recognized as update in layouter and for every update action on component, antoher update is triggerd
  highlight: string | undefined = undefined;
  type: UMLElementType = PetriNetElementType.PetriNetPlace;
  bounds: IBoundary = { ...this.bounds, width: 60, height: 60 };
  amountOfTokens: number;
  capacity: number;

  constructor(values?: DeepPartial<UMLPetriNetPlace>) {
    super(values);
    assign<IUMLElement>(this, values);
    this.amountOfTokens = values?.amountOfTokens || values?.amountOfTokens === 0 ? values.amountOfTokens : 0;
    this.capacity = values?.capacity || values?.capacity === 0 ? values.capacity : UMLPetriNetPlace.defaultCapacity;
  }

  serialize(children?: UMLElement[]): Apollon.UMLPetriNetPlace {
    return {
      ...super.serialize(),
      type: this.type as keyof typeof PetriNetElementType,
      amountOfTokens: this.amountOfTokens,
      capacity: !isFinite(this.capacity) ? this.capacity.toString() : this.capacity,
    };
  }

  deserialize<T extends Apollon.UMLModelElement>(values: T, children?: Apollon.UMLModelElement[]) {
    const assert = (v: Apollon.UMLModelElement): v is Apollon.UMLPetriNetPlace =>
      v.type === PetriNetElementType.PetriNetPlace;
    if (!assert(values)) {
      return;
    }

    super.deserialize(values, children);
    this.amountOfTokens = values.amountOfTokens;
    this.capacity =
      values.capacity === Number.POSITIVE_INFINITY.toString() ? Number.POSITIVE_INFINITY : (values.capacity as number);
  }

  render(canvas: ILayer): ILayoutable[] {
    return [this];
  }
}
