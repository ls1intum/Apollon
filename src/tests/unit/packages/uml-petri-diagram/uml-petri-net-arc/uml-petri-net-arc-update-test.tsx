import React from 'react';
import { UMLElement } from '../../../../../main/services/uml-element/uml-element';
import { UMLClassBidirectional } from '../../../../../main/packages/uml-class-diagram/uml-class-bidirectional/uml-class-bidirectional';
import { Direction } from '../../../../../main/services/uml-element/uml-element-port';
import { getRealStore } from '../../../test-utils/test-utils';
import { wrappedRender } from '../../../test-utils/render';
import { fireEvent } from '@testing-library/react';
import { UMLPetriNetPlace } from '../../../../../main/packages/uml-petri-net/uml-petri-net-place/uml-petri-net-place';
import { UMLPetriNetArc } from '../../../../../main/packages/uml-petri-net/uml-petri-net-arc/uml-petri-net-arc';
import { UMLPetriNetArcUpdate } from '../../../../../main/packages/uml-petri-net/uml-petri-net-arc/uml-petri-net-arc-update';

describe('test petri net arc update', () => {
  let elements: UMLElement[] = [];
  let source: UMLPetriNetPlace;
  let target: UMLPetriNetPlace;
  let petriNetArc: UMLPetriNetArc;

  beforeEach(() => {
    // initialize  objects
    source = new UMLPetriNetPlace({ id: 'source-test-id' });
    target = new UMLPetriNetPlace({ id: 'target-test-id' });
    petriNetArc = new UMLPetriNetArc({
      id: 'test-id',
      name: 'UMLClassBidirectional',
      source: { element: source.id, direction: Direction.Up },
      target: { element: target.id, direction: Direction.Up },
    });
    elements.push(source, target, petriNetArc);
  });

  it('render', () => {
    const store = getRealStore(undefined, elements);

    const { baseElement } = wrappedRender(<UMLPetriNetArcUpdate element={petriNetArc} />, { store });
    expect(baseElement).toMatchSnapshot();
  });

  it('flip', () => {
    const store = getRealStore(undefined, elements);

    const { getAllByRole } = wrappedRender(<UMLPetriNetArcUpdate element={petriNetArc} />, { store: store });
    const buttons = getAllByRole('button');
    buttons[0].click();

    const element = store.getState().elements[petriNetArc.id] as UMLClassBidirectional;

    expect(element.target).toEqual(petriNetArc.source);
    expect(element.source).toEqual(petriNetArc.target);
  });

  it('delete', () => {
    const store = getRealStore(undefined, elements);

    const { getAllByRole } = wrappedRender(<UMLPetriNetArcUpdate element={petriNetArc} />, { store: store });
    const buttons = getAllByRole('button');
    buttons[1].click();

    expect(store.getState().elements).not.toContain(petriNetArc.id);
  });

  it('change name', () => {
    const store = getRealStore(undefined, elements);

    const { getByRole } = wrappedRender(<UMLPetriNetArcUpdate element={petriNetArc} />, {
      store: store,
    });
    const nameField = getByRole('textbox');
    const updatedValue = '1';
    fireEvent.change(nameField, { target: { value: updatedValue } });

    const updatedElement = store.getState().elements[petriNetArc.id] as UMLPetriNetArc;

    expect(updatedElement.name).toEqual(updatedValue);
  });
});
