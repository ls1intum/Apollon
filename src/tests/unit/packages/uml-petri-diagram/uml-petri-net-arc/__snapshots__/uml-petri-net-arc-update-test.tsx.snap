// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test petri net arc update render 1`] = `
<body>
  <div>
    <div>
      <section>
        <div
          class="sc-ehmTmK deAjaH"
        >
          <input
            class="sc-ezWOiH sc-bZkfAO wLKMs gTZQoR"
            value="UMLClassBidirectional"
          />
          <button
            class="sc-fLlhyt sc-bBrHrO isqEak bBoMhO"
            color="link"
          >
            <svg
              class="sc-jqUVSM iZiiCQ"
              height="16px"
              viewBox="0 0 512 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M508.485 184.485l-92.485 92c-4.687 4.686-12.284 4.686-16.97 0l-7.071-7.07c-4.687-4.686-4.687-12.284 0-16.971L452.893 192H12c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h440.905l-60.946-60.444c-4.687-4.686-4.687-12.284 0-16.971l7.07-7.07c4.687-4.686 12.284-4.686 16.971 0l92.485 92c4.687 4.686 4.686 12.284 0 16.97zm-504.97 160l92.485 92c4.687 4.686 12.284 4.686 16.971 0l7.07-7.07c4.687-4.686 4.687-12.284 0-16.971L59.095 352H500c6.627 0 12-5.373 12-12v-8c0-6.627-5.373-12-12-12H59.107l60.934-60.444c4.687-4.686 4.687-12.284 0-16.971l-7.071-7.07c-4.686-4.686-12.284-4.687-16.97 0l-92.485 92c-4.686 4.686-4.687 12.284 0 16.97z"
              />
            </svg>
          </button>
          <button
            class="sc-fLlhyt sc-bBrHrO isqEak bBoMhO"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-jqUVSM iZiiCQ"
              height="16px"
              viewBox="0 0 448 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
              />
            </svg>
          </button>
        </div>
      </section>
    </div>
  </div>
</body>
`;
